src/utils/formatNumber.ts

```diff
@@ -137,3 +137,57 @@ export const formatMinAmount = (minAmount: sdkAmount.Amount): string => {
     return asNumber.toPrecision(3);
   }
 };
+
+/**
+ * Format a number with smart precision based on total digits and max decimal digits.
+ * If the integer part is less than totalDigits, show decimals up to maxDecimals
+ * (limited by remaining space from totalDigits).
+ *
+ * @param value - The numeric value to format (string or number)
+ * @param totalDigits - Maximum total significant digits to display
+ * @param maxDecimals - Maximum number of decimal places to show
+ * @returns Formatted number string (decimals truncated)
+ *
+ * @example
+ * formatWithPrecision('123.456789', 6, 4)     // '123.456' (3 int + 3 dec = 6 total)
+ * formatWithPrecision('123.456789', 8, 4)     // '123.4567' (3 int + 4 dec, limited by maxDecimals)
+ * formatWithPrecision('123456.789', 6, 4)     // '123456' (6 int, no space for decimals)
+ * formatWithPrecision('1234567.89', 6, 4)     // '1234567' (int exceeds totalDigits, show full int)
+ */
+export const formatWithPrecision = (
+  value: string | number,
+  totalDigits: number,
+  maxDecimals: number,
+): string => {
+  const numValue = typeof value === 'string' ? parseFloat(value) : value;
+
+  // Handle zero and NaN cases
+  if (numValue === 0 || Number.isNaN(numValue)) {
+    return '0';
+  }
+
+  // Note: We don't use toPrecision() because:
+  // 1. It rounds instead of truncates
+  // 2. It doesn't respect maxDecimals constraint properly
+  // 3. It can return scientific notation for large/small numbers
+
+  // Get the integer part length (count digits, not including decimal)
+  const integerPart = Math.floor(numValue);
+  const integerDigits = integerPart === 0 ? 1 : integerPart.toString().length;
+
+  // If integer part already exceeds or equals totalDigits, return just the integer
+  if (integerDigits >= totalDigits) {
+    return integerPart.toString();
+  }
+
+  // Calculate how many decimal places we can show
+  const remainingDigits = totalDigits - integerDigits;
+  const decimalsToShow = Math.min(remainingDigits, maxDecimals);
+
+  // Truncate decimals to match the max allowed
+  const [intPart, decimalPart] = numValue.toString().split('.');
+  if (!decimalPart || decimalsToShow === 0) {
+    return intPart;
+  }
+  return `${intPart}.${decimalPart.substring(0, decimalsToShow)}`;
```

Comment:
Guessing the Amount utils dont help us with anything here?
Like truncate and display https://github.com/wormhole-foundation/wormhole-sdk-ts/blob/30810950efb84e3244d8e6c6f11d349b1e046deb/core/base/src/utils/amount.ts#L68